### DISCUSSION TITLE
Which mode is the canonical for typescript in combination with eglot?

### DISCUSSION THREAD

**Discussion by HaraldKi** [joaotavora/eglot#1378] [Q&A] [answered]
Eglot relies on `eglot-server-programs` to trigger a server based on language mode. There is `typescript-mode`, `js-ts-mode`, `typescript-ts-mode` and more. Yet when I try to figure out which packages provide these modes, I find packages providding "language parsing" features (treesitter, tide), while I thought the language parsing is left to the language server.

Is there a "canonical" typescript mode meanwhile or am I misunderstanding this. At least quite some people seem to combine treesitter with eglot. Would it work to use a completely empty `my-ts-mode` (I should really try this soon :-) and trigger the language server?

**Comment by HaraldKi** [comment]
This gets weirder by the minute. I understand so far:
- tree-sitter does not even have language grammars (sounds right)
- but every place on the net suggests to install some or other language grammar for tree-sitter
I maintain that this is odd in the context of eglot, as I would expect that the language server is responsible for providing the grammar. So I did this:
```elisp
(define-derived-mode hk-typescript-mode prog-mode "HK-TS" ()
  "Marks a file as typescript to get eglot going."
  )
...
(add-to-list
 'eglot-server-programs
 `(hk-typescript-mode
   "typescript-language-server" "--stdio"
   :initializationOptions
   (:tsserver (:path "./node_modules/typescript/lib/tsserver.js"))))
```
Further I have
```elisp
(with-eval-after-load 'eglot
  (define-key eglot-mode-map (kbd "M-Q") 'eglot-format-buffer)
  (define-key eglot-mode-map (kbd "M-/") 'company-complete)
  (define-key eglot-mode-map (kbd "C-\\") 'eglot-code-actions)
...
```
and this starts to work. **I don't get syntax highlighting yet** (and get the impression it is not supported??) and there may be other things missing.

But my assumption seems to be correct: if I can convince eglot to take over with a major-mode->language-server mapping and wire my preferred key-bindings, things get going. No other make-shift (sorry) language parsing necessary. And really, this is exactly how things should work and I am really happy with this slick and clean approach: the language server knows the language and the editor does the editing.

Let me know if I am missing something important that tree-sitter and/or those parsers for it would add.

**Reply by HaraldKi** [reply]
Uff, no Semantic Tokens handled by eglot.el :frowning_face:

**Reply by joaotavora** [reply]
> Uff, no Semantic Tokens handled by eglot.el ☹️

Yup, because even if these "semantic tokens" were handled by Eglot, there is no machinery in place to provide indentation and fontification support efficiently.  

Treesitter + grammars do this analysis in the same process address space where Emacs buffers live LSP do it necessarily in a different address space.  

This is almost surely less efficient (but it remains to be tested). 
This is almost surely hard to implement (judging by how hard it was to implement tree-sitter).

**Reply by HaraldKi** [reply]
Not sure what "hard to implement" exactly refers to: technically at all or technically and efficient?

Because I want to try this anyway, taking inspiration and hints from lsp-mode. But don't wait for it. I know Java and Typescript quite well, but in elisp I have to look up everything. So the code, even if it works, will probably not be what one wants. Let's see how far I get. :smile: I'll consider it as an interesting exercise.

**Reply by joaotavora** [reply, marked as answer]
Support for LSP Semantic Token highlighting is now in Eglot master branch.  You can try it out by compiling the latest master branch of Emacs, or waiting a few hours and grabbing te ELPA package from GNU-devel ELPA.  https://elpa.gnu.org/devel/eglot.html.  Give it as much testing as possible.  The feature will be in Eglot 1.20.
